bazel/utils/remote.bzl

```diff
@@ -0,0 +1,300 @@
+    "template": attr.label(
+        default = "@enkit//bazel/utils/remote:runner.template.sh",
+        allow_single_file = True,
+        doc = "template to use to generate the shell script to run the target remotely",
+    ),
+    "tools": attr.label_list(
+        allow_files = True,
+        cfg = "host",
+        doc = "Additional tools to require on YOUR host to perform the copy. " +
+              "This is useful - for example - to use a custom rsync or ssh binary (specify a bazel relative path as ssh_bin, referencing the binary generated by the included target)",
+    ),
+}
+
+def _common_attrs_to_dict(ctx):
+    """Converts all the common attributes into a dictionary."""
+    attrs = {}
+    for name in _common_attrs:
+        value = getattr(ctx.attr, name, None)
+        if value == None:
+            continue
+        attrs[name] = value
+
+    return attrs
+
+def _remote_run_impl(ctx):
+    has_wrapper = ctx.attr.wrapper and not package(ctx.attr.wrapper.label) == package(ctx.attr.noop.label)
+
+    attrs = _common_attrs_to_dict(ctx)
+    if has_wrapper and RemoteWrapper in ctx.attr.wrapper:
+        attrs = merge_kwargs(attrs, ctx.attr.wrapper[RemoteWrapper].attributes)
+
+    attrs = struct(**attrs)
+    if package(attrs.target.label) == package(attrs.noop.label):
+        fail(location(ctx) + "A target must be supplied via flags - read the file '//" + ctx.build_file_path + "' for details")
+
+    target_exec = attrs.target[DefaultInfo].files_to_run.executable
+    target_runfiles = attrs.target[DefaultInfo].default_runfiles
+    target_opts = attrs.target_opts
+
+    runfiles = ctx.runfiles(files = [target_exec])
+    runfiles = runfiles.merge(target_runfiles)
+
+    if has_wrapper:
+        wrapper_exec = ctx.attr.wrapper[DefaultInfo].files_to_run.executable
+        wrapper_runfiles = ctx.attr.wrapper[DefaultInfo].default_runfiles
+        runfiles = runfiles.merge(wrapper_runfiles)
+        runfiles = runfiles.merge(ctx.runfiles(files = [wrapper_exec]))
+
+        target_opts = attrs.wrapper_opts + ["--", target_exec] + target_opts
+        target_exec = wrapper_exec
+
+    include = ctx.outputs.include
+    ctx.actions.write(include, "\n".join([ctx.workspace_name + "/" + f.short_path for f in runfiles.files.to_list()]))
+
+    subs = dict(
+        include = shell.quote(include.short_path),
+        destdir = shell.quote(attrs.destdir),
+        target = shell.quote(package(attrs.target.label)),
+        target_opts = shell.array_literal(attrs.target_opts),
+        executable = shell.quote(target_exec.short_path),
+        workspace = shell.quote(ctx.workspace_name),
+        rsync_opts = shell.array_literal(attrs.rsync_opts),
+        rsync_bin = shell.quote(attrs.rsync_bin),
+        ssh_opts = shell.array_literal(attrs.ssh_opts),
+        ssh_bin = shell.quote(attrs.ssh_bin),
+        machines = shell.array_literal(attrs.machines),
```

Comment:
Is there some way we can make this machine list available to the runner target as an environment variable / argument? The first machine takes on the execution role, but it would need access to a host list when running distributed jobs.

Reply:
added, great idea.